node/pkg/watchers/solana/shim.go

```diff
@@ -0,0 +1,362 @@
+
+package solana
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/certusone/wormhole/node/pkg/common"
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"github.com/near/borsh-go"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const (
+	shimPostMessageDiscriminatorStr  = "d63264d12622074c"
+	shimMessageEventDiscriminatorStr = "e445a52e51cb9a1d441b8f004d4c8970"
+)
+
+type (
+
+	// ShimPostMessageData defines the shim PostMessage payload following the eight byte discriminator (shimPostMessageDiscriminatorStr)
+	ShimPostMessageData struct {
+		Nonce            uint32
+		ConsistencyLevel ConsistencyLevel
+		Payload          []byte
+	}
+
+	// ShimMessageEventData defines the shim MessageEvent payload following the sixteen byte discriminator (shimMessageEventDiscriminatorStr)
+	ShimMessageEventData struct {
+		EmitterAddress [32]byte
+		Sequence       uint64
+		Timestamp      uint32
+	}
+
+	// ShimAlreadyProcessed is a map that tracks the inner index identifier pairs already processed as part of shim transactions.
+	// This allows us to avoid double processing instructions.
+	ShimAlreadyProcessed map[ShimInnerIdx]struct{}
+
+	// InnerIdx is the key to the set of instructions already processed as part of shim transactions
+	ShimInnerIdx struct {
+		topLevelIdx int
+		innerIdx    int
+	}
+)
+
+// add adds the specified inner index identifier pair into a ShimAlreadyProcessed.
+func (sap *ShimAlreadyProcessed) add(outerIdx int, innerIdx int) {
+	(*sap)[ShimInnerIdx{outerIdx, innerIdx}] = struct{}{}
+}
+
+// exists returns true if the specified inner index identifier pair is in a ShimAlreadyProcessed.
+func (sap *ShimAlreadyProcessed) exists(outerIdx int, innerIdx int) bool {
+	_, exists := (*sap)[ShimInnerIdx{outerIdx, innerIdx}]
+	return exists
+}
+
+// shimSetup performs any initialization that is specific to monitoring the shim contract.
+func (s *SolanaWatcher) shimSetup() {
+	s.shimEnabled = s.shimContractStr != ""
+	if s.shimEnabled {
+		var err error
+		s.shimPostMessageDiscriminator, err = hex.DecodeString(shimPostMessageDiscriminatorStr)
+		if err != nil {
+			panic("failed to decode shim post message discriminator")
+		}
+		s.shimMessageEventDiscriminator, err = hex.DecodeString(shimMessageEventDiscriminatorStr)
+		if err != nil {
+			panic("failed to decode shim post message discriminator")
+		}
+	}
+}
+
+// shimMatchPrefix verifies that the instruction data starts with the specified prefix bytes.
+func shimMatchPrefix(discriminator []byte, buf []byte) bool {
+	if len(buf) < len(discriminator) {
+		return false
+	}
+	return bytes.Equal(discriminator, buf[:len(discriminator)])
+}
+
+// shimParsePostMessage parses a shim PostMessage and returns the results.
+func shimParsePostMessage(shimPostMessageDiscriminator []byte, buf []byte) (*ShimPostMessageData, error) {
+	if len(buf) <= len(shimPostMessageDiscriminator) {
+		return nil, errors.New("payload too short")
```

Comment:
do we need this check (and the other "payload too short" one) if `shimMatchPrefix` performs the length check?